_posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd

```diff
@@ -0,0 +1,753 @@
+
+Secondly, this post can serve as an introduction to building your own modules. This is something you may be familiar with from Keras or not -- depending on whether you've used custom models or rather, preferred the declarative *define -\> compile -\> fit* style. (Yes, I'm implying there's some transfer going on if one comes to `torch` from Keras custom training. Syntactic and semantic details may be different, but both share the object-oriented style that allows for great flexibility and control.)
+
+Last not least, we'll also use this as a hands-on experience with RNN architectures (the LSTM, specifically). While the general concept of recurrence may be easy to grasp, it is not necessarily self-evident how those architectures should, or could, be coded. Personally, I find that independently of the framework used, RNN-related documentation leaves me confused. What exactly *is* being returned from calling an LSTM, or a GRU? (In Keras this depends on how you've defined the layer in question.) I suspect that once we've decided what we *want* to return, the actual code won't be that complicated. Consequently, it should be worth our time to take a detour clarifying what it is that `torch` and Keras are giving us. Implementing our convLSTM will be a lot more straightforward thereafter.
+
+## A `torch` convLSTM
+
+The code discussed here may be found on [Github](https://github.com/skeydan/convlstm). (Depending on when you're reading this, the code in that repository may have evolved though.)
+
+My starting point was one of the PyTorch implementations found on the net, namely, [this one](https://github.com/ndrplz/ConvLSTM_pytorch/blob/master/convlstm.py). If you search for "PyTorch convGRU" or "PyTorch convLSTM", you will find stunning discrepancies in how these are realized -- discrepancies not just in syntax and/or engineering ambition, but on the semantic level, right at the center of what the architectures may be expected to do. As they say, let the buyer beware. (Regarding the realization I ended up porting, I am confident that while numerous optimizations will be possible, the basic mechanism matches my expectations.
+
+What do I expect? Let's approach this task in a top-down way.
+
+### Input and output
+
+The convLSTM's input will be a time series of spatial data, each observation being of size `(time steps, channels, height, width)`.
+
+Compare this with the usual RNN input format, be it in `torch` or Keras. In both frameworks, RNNs expect tensors of size `(timesteps, input_dim)`[^1]. `input_dim` is $1$ for univariate time series and greater than $1$ for multivariate ones. Conceptually, we may match this to convLSTM's `channels` dimension: There could be a single channel, for temperature, say -- or there could be several, such as for pressure, temperature, and humidity. The two additional dimensions found in convLSTM, `height` and `width`, are spatial indexes into the data.
```

Comment:
lol :-) I tried again ;-) ... I really prefer `indexes` (but that may be because I had several years (((working as a DBA))) to digest it ;-)))

Re technical language

> Meanwhile, indices is generally preferred in mathematical, financial, and technical contexts, while indexes is relatively common in general usage.

I'm not sure how they mean it, but personally I'd only ever use indices when referring to i, j, k or the like, but otherwise ... like in the above sentence ... I think indexes sounds more natural ...